sno/conflicts.py

```diff
@@ -171,58 +172,264 @@ def _ensure_conflict(self, conflict):
+    But if renames have occurred, it could have multiple names, eg:
+    "ancestor=table_A:fid=5 ours=table_A:fid=6 theirs=table_A:fid=12"
+    """
+    dpath3 = decoded_path3
+    actual_dpaths = [p for p in dpath3 if p]
+    all_same_path = len(set(actual_dpaths)) == 1
+
+    if all_same_path:
+        return decoded_path_to_label(actual_dpaths[0])
+
+    label3 = AncestorOursTheirs(
+        *(
+            f"{v}={decoded_path_to_label(p)}" if p else None
+            for v, p, in zip(AncestorOursTheirs.NAMES, decoded_path3)
+        )
+    )
+    return " ".join([l for l in label3 if l])
+
+
+def decoded_path_to_label(decoded_path):
+    """
+    Converts a decoded path to a unique name, eg:
+    ("table_A", "feature", "fid", 5) -> "table_A:fid=5"
+    """
+    if decoded_path is None:
+        return None
+    if decoded_path[1] == "feature":
+        table, tablepart, pk_field, pk = decoded_path
+        return f"{table}:{pk_field}={pk}"
+    else:
+        return ":".join(decoded_path)
+
+
+def add_conflict_dict_to_category(root_dict, conflict_category, conflict_dict):
+    """
+    Ensures the given category of conflicts exists, and then adds
+    the given conflict dict to it.
+    """
+    cur_dict = root_dict
+    for c in conflict_category:
+        cur_dict.setdefault(c, {})
+        cur_dict = cur_dict[c]
+
+    cur_dict.update(conflict_dict)
+
+
+def summarise_conflicts(cur_dict, output_format):
+    """
+    Recursively traverses the tree of categorised conflicts,
+    looking for a dict where the values are placeholders.
+    For example:
+    {
+        K1: _CONFLICT_PLACEHOLDER,
+        K2: _CONFLICT_PLACEHOLDER,
+    }
+    When found, it will be replaced with one of the following:
+    1) SHORT_SUMMARY: 2 (the size of the dict)
+    2) SUMMARY: [K1, K2]
+    """
+    first_value = next(iter(cur_dict.values()))
+    if first_value == _CONFLICT_PLACEHOLDER:
+        if output_format == ConflictOutputFormat.SHORT_SUMMARY:
+            return len(cur_dict)
+        else:
+
+            def sortkey(item):
+                tableybit, pk = item.split('=')
+                pk = int(pk) if pk.isdigit() else pk
+                return tableybit, pk
+
+            return sorted([k for k in cur_dict.keys()], key=_label_sort_key)
+
+    for k, v in cur_dict.items():
+        cur_dict[k] = summarise_conflicts(v, output_format)
+    return cur_dict
+
+
+def _label_sort_key(label):
+    """Sort labels of conflicts in a sensible way."""
+    if (
+        label.startswith("ancestor=")
+        or label.startswith("ours=")
+        or label.startswith("theirs=")
+    ):
+        # Put the complicated conflicts last.
+        return "Z multiple-path", label
+
+    parts = label.split('=', 1)
+    if len(parts) == 2:
+        prefix, pk = parts
+        pk = int(pk) if pk.isdigit() else pk
+        return "Feature", prefix, pk
+    else:
+        # TODO: maybe meta conflicts should go before feature conflicts.
```

Comment:
any reason not to just do that now? makes sense

Reply:
Done